packages/twoside/init.lua

```diff
@@ -65,28 +66,74 @@ function package:_init (options)
   self.class:registerPostinit(function (class)
     class:mirrorMaster(options.oddPageMaster, options.evenPageMaster)
   end)
+  self.class:registerHook("newpage", spreadHook)
   self.class:registerHook("newpage", switchPage)
 end
 
 function package:registerCommands ()
 
   self:registerCommand("open-double-page", function()
-    SILE.typesetter:leaveHmode()
+    SILE.call("open-spread", { double = false, odd = true, blank = false })
+  end)
+
+  self:registerCommand("open-spread-eject", function()
     SILE.call("supereject")
-    if self.class:oddPage() then
-      SILE.typesetter:typeset("")
-      SILE.typesetter:leaveHmode()
-      SILE.call("supereject")
+  end)
+
+  -- This is upstreamed from CaSILE. Similar to the original open-double-page,
+  -- but can disable headers and folios on blank pages and allows opening the
+  -- even side (with or without a leading blank).
+  self:registerCommand("open-spread", function (options)
+    local odd = SU.boolean(options.odd, true)
+    local double = SU.boolean(options.double, true)
+    local blank = SU.boolean(options.blank, true)
+    local optionsMet = function ()
+      return (not double or spread_counter > 1) and
+             (odd and self.class:oddPage()) or
+             (not odd and not self.class:oddPage())
     end
+    spread_counter = 0
     SILE.typesetter:leaveHmode()
+    -- Output a box, then remove it otherwise we can't prove what page new content will land on
+    SILE.call("hbox"); SILE.typesetter:leaveHmode(); table.remove(SILE.typesetter.state.nodes)
+    if spread_counter == 1 and optionsMet() then
+      SILE.typesetter.state.outputQueue = {}
+      return
+    end
+    repeat
+      if spread_counter > 0 then
+        SILE.call("hbox")
+        SILE.typesetter:leaveHmode()
+        if blank then
+          SILE.scratch.headers.skipthispage = true
+          SILE.scratch.counters.folio.off = 2
+        end
+      end
+      SILE.call("open-spread-eject")
+      SILE.typesetter:leaveHmode()
+    until optionsMet()
   end)
 
 end
 
 package.documentation = [[
 \begin{document}
 The \code{book} class described in chapter 4 sets up left and right mirrored page masters; the \autodoc:package{twoside} package is responsible for swapping between the two left and right frames, running headers and so on.
-It has no user-serviceable parts.
+It's main function in mirroring master frames does not provide any user-serviceable parts.
+It does supply a couple user facing commands for convenince.
+
+The \autodoc:command{\open-double-page} ejects whatever page is currently being processed, then checks if it landed on an odd page.
```

Comment:
Regarding the discussion whether to put this in a book class, but the lack of class documentation mechanism:
Why not have these commands in a standalone package, loaded by the book class. This way it could have documentation, and I guess be used even if twoside is not active?

Reply:
Because I'm trying **not** to refactor anything that would be a breaking change here, I'm trying to fix a bug so I can publish a book with chapter headings on odd pages.

I 100% agree refactoring this elsewhere makes sense, but I don't have the bandwidth to figure out whether the class or another package makes more sense or how to do that without it being a breaking change. Since it was already in this file and the version I'm submitting has bugs fixed and gets things on the right pages (which the original does not) I'm going with it for now.